guides/source/threading_and_code_execution.md

```diff
@@ -5,65 +5,145 @@ Threading and Code Execution in Rails
-As long as your code doesn't modify any of those shared things, it can mostly
-ignore that other threads exist.
+Asynchronous Active Record queries are also performed in the background,
+allowing other processes to run on the main thread.
 
-The rest of this guide describes the mechanisms Rails uses to make it "mostly
-ignorable", and how extensions and applications with special needs can use them.
+The above mechanisms all involve multiple threads, each managing work for a
+unique instance of an object (controller, job, channel), while sharing the
+global process space (such as classes and their configurations, and global
+variables). As long as the code on each thread doesn't modify anything shared,
+multiple threads can safely run concurrently.
 
-Executor
---------
+Rails' built-in concurrency will cover the day-to-day needs of most application
+developers, and ensure applications remain generally performant.
 
-The Rails Executor separates application code from framework code: any time the
-framework invokes code you've written in your application, it will be wrapped by
-the Executor.
+NOTE: You can read more about how to configure Rails' concurrency in the
+[Framework Behavior](#framework-behavior) section.
 
-The Executor consists of two callbacks: `to_run` and `to_complete`. The Run
-callback is called before the application code, and the Complete callback is
-called after.
 
-### Default Callbacks
+### Isolated Execution State
 
-In a default Rails application, the Executor callbacks are used to:
+The `active_support.isolation_level` value in your `configuration.rb` file
+provides you the option to define where Rails' internal state should be stored
+while tasks are run. If you use a fiber-based server or job processor (e.g.
+[`falcon`](https://github.com/socketry/falcon)), you should set this value to
+`:fiber`, otherwise it is best to set it to `:thread`.
 
-* track which threads are in safe positions for autoloading and reloading
-* enable and disable the Active Record query cache
-* return acquired Active Record connections to the pool
-* constrain internal cache lifetimes
+Later sections of this guide detail advanced ways of wrapping code to ensure
+thread safety, and how extensions and applications with particular concurrency
+requirements, such as library maintainers, should do this.
 
-Prior to Rails 5.0, some of these were handled by separate Rack middleware
-classes (such as `ActiveRecord::ConnectionAdapters::ConnectionManagement`), or
-directly wrapping code with methods like
-`ActiveRecord::Base.connection_pool.with_connection`. The Executor replaces
-these with a single more abstract interface.
+Storing Thread-specific Data
+----------------------------
 
-### Wrapping Application Code
+The
+[`ActiveSupport::CurrentAttributes`](https://edgeapi.rubyonrails.org/classes/ActiveSupport/CurrentAttributes.html)
+class is a special class in Rails that helps you manage temporary,
```

Comment:
Fussing over a detail after I've suggested this section might change radically, but "The X class is a special class in Rails that helps you Y" feels like a rather long walk (and "special" is unclear).